### PR TITLE
[mergify] backport automation to 7.x, 7.12 and 7.11

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

Enable [backports](https://docs.mergify.io/actions/backport.html) with [mergify](https://docs.mergify.io/index.html).

## Why is it important?

`Mergify` provides different mechanism to interact with PRs and automate certain process.

## Use cases

We enabled it for the  https://github.com/elastic/apm-pipeline-library/ project for three different automations:
* Automatically merge to master  ( if >2 approvals, CLA is ✅ , label=ready-to-merge and CI ✅ )
* Automatically backport to 7.x (if origin=master, label=backport-to-7.x)
* Merge backports ( if CLA is ✅  and CI ✅)

### Some examples

Backports to 7.x -> https://github.com/elastic/apm-pipeline-library/pull/995
Merges to master -> https://github.com/elastic/apm-pipeline-library/pull/994

## Pros

- It does not require any credentials.
- It works out of the box.
- It provides GitHub [comments](https://docs.mergify.io/commands.html).
- Extend functionality with more automation.
- `infosec` team approved.

## Cons

- It's a GitHub APP so potentially the service could be deprecated.
  - **[mitigation]** -> Nothing we can do about it. 
- It does not use the `.backportrc.json` 
  - **[mitigation]** -> .backportrc.json` is automatically modified when new branches are created -> https://github.com/elastic/beats/pull/24098 . Therefore `.mergify.yml` could be also modified accordingly.
- Backport are not assigned to the original contributor.
  - **[mitigation]** -> To explore if there is kind of way to do so.
- Only free for open repos. See comment https://github.com/elastic/beats/pull/24608#issuecomment-803873649
  - **[mitigation]** -> To explore if private repos are a thing and if so whether the payment plan is sufficient.

## Issues

* Duplicates https://github.com/elastic/beats/pull/24576
* There is an initiative to automate the GitHub labels creation in each obs11 repo -> https://github.com/elastic/apm-pipeline-library/pull/886

## Actions

- [x] Ask Infra to enable the `mergify` app in this project.
- [x] Agree with the Beats team
- [x] Document the process. I documented the process in the internal repositories instead of  https://github.com/elastic/beats/tree/master/docs/devguide (thanks @cachedout)

### CODE REVIEW

**Review by kuisathaverat — APPROVED** [review]

File: .mergify.yml
Comment (sqren) [reply .mergify.yml:5]: We should use the existing labels like `v7.13.0` so people don't have to duplicate them.

**Review by cachedout — APPROVED** [review]
I know this is a backport (sorry!) and I should have commented on the original,  but I didn't realize until now -- did the original come with any developer documentation? 

It might be good to find a home for this somewhere in here: https://github.com/elastic/beats/tree/master/docs/devguide

**Comment by ruflin** [comment]
One thing that becomes more and more important from my perspective is that we have consistency across repositories. There are quite a few engineers that work on beats, kibana, integrations and other repositories. Having to use different tools for each repository is a hurdle. The same is the case for the labels, they have to be consistent.

Kibana uses the backport tool today and it seems to work well. So I would like to understand why we can't do the same here as Kibana.

**Comment by v1v** [comment]
In order to help with the rationale about this PR, let me highlight what happened for having two different PRs for a similar problem:

1) https://github.com/elastic/beats/pull/24576 was raised.
2) there was a conversation regarding the implications for using GitHub actions with credentials.
3) productivity team gathered to discuss what to do.
  a) for clarity this requirement was already tracked in our TODO list but no yet prioritised.
  b) this requirement could be related to the `merge-queue` proposal for the beats project, that was intentionally postponed until 7.15, when the changelog could be improved.
4) We came up with this proposal and explained the PROS/CONS.

In a nutshell:
- in both PRs the aim is exactly the same for the end user. 
  - add a GitHub label and the the automation will happen
- the process will remain the same from the user point of view.
- the only differences are the ones related to the implementation details.

File: .mergify.yml
Comment (v1v) [reply .mergify.yml:5]: Yes, I added a comment to remember we need to set the labels accordingly.

For the record, we have a WiP automation to apply the same labels across all the observability repositories, therefore it will be quite simple to apply a new label everywhere.

**Comment by ruflin** [comment]
> in both PRs the aim is exactly the same for the end user

This sounds great. As long as the dev does not need to have specific knowledge if it is in Kibana or Beats, this sounds great. And I think short term it is more important to having A tool than none.

I'm pretty sure the devil / differences are somewhere in the details but we will find out. Lets keep moving on this in case @sqren agrees. My main ask is to make sure labels are aligned so the statement above holds true.

Long term I think we should align tooling.

**Comment by sqren** [comment]
Going the mergify route is good with me. I agree with @ruflin that the user experience should be the same if you contribute to kibana or beats.

I'm curious to hear what happens when the mergify backport fails. How does it delegate the resolution to the end user? 
The Github action we use in kibana will let the user know how to run the backport command locally:

image

Running `node scripts/backport --pr 94072` will then auto-select the remaining branches, cherry-pick the changes. When the user has resolved the conflicts for each branch the tool will create the PRs as usual.

**Comment by v1v** [comment]
> I'm curious to hear what happens when the mergify backport fails. How does it delegate the resolution to the end user?

The backport action in `mergify` creates the PRs with conflicts by default and tag them with a specific label, further details in  
https://docs.mergify.io/actions/backport.html

It won't add any notifications to the original PR. But create a PR with the label. See below to illustrate a real example:

https://github.com/elastic/apm-pipeline-library/pull/978 is one example for instance

image

>The Github action we use in kibana will let the user know how to run the backport command locally

`mergify` supports GitHub comments to interact with the `mergify` actions, therefore no need to run anything by command line at first. For instance, see the below snippet that I took from the above PR.

> More conditions and actions can be found in the documentation.
You can also trigger Mergify actions by commenting on this pull request:
@Mergifyio refresh will re-evaluate the rules
@Mergifyio rebase will rebase this PR on its base branch
@Mergifyio update will merge the base branch into this PR
@Mergifyio backport <destination> will backport this PR on <destination> branch


Further details in https://docs.mergify.io/commands.html

**Comment by sqren** [comment]
> therefore no need to run anything by command line at first. 

The user only need to run the backport tool locally if a conflict occurs. Similar to how the mergify approach requires the user to checkout the branch locally and resolve the conflicts. imo it could be a little more tedious, especially if there are multiple branches with conflicts (like the example above) but we'll see.

**Comment by kuisathaverat** [comment]
we can add a comment with the instructions to backport the PR with backport locally in case of failure https://docs.mergify.io/examples.html#welcoming-your-contributors

**Comment by sqren** [comment]
> we can add a comment with the instructions to backport the PR with backport locally in case of failure [docs.mergify.io/examples.html#request-for-action](https://docs.mergify.io/examples.html#request-for-action)

That sounds like a good idea 👍

File: .mergify.yml
Comment (v1v) [inline .mergify.yml:8]: `author` won't be resolved correctly -> https://github.com/Mergifyio/mergify-engine/issues/1096

Comment (v1v) [reply .mergify.yml:8]: See https://github.com/v1v/apm-pipeline-library/pull/19#issuecomment-802306489

image

**Comment by cachedout** [comment]
> Long term I think we should align tooling.

Just want to mention here that while Mergify is free for public repos (and very good IMHO!), its [per-organization pricing ](https://mergify.io/pricing)may be a barrier if we ever want to roll it out to private repos. If tooling consistency across repos/teams is a concern, we should take this into account.

File: .mergify.yml
Comment (urso) [reply .mergify.yml]: > I think we should use the existing labels like v7.13.0 so people don't have to duplicate them.

We backport too often after FF, sometimes right before release time leaving us with fixes that didn't make it into the release. From past experience we've been notoriously bad at telling which exact a fix/features is in. I think I would prefer to tell the target branch instead of the exact version.

**Comment by urso** [comment]
> The user only need to run the backport tool locally if a conflict occurs. Similar to how the mergify approach requires the user to checkout the branch locally and resolve the conflicts. imo the mergify approach is a little more tedious, especially if there are multiple branches with conflicts (like the example above) but we'll see.

Master and 7.x are mostly in sync. We seldomly have conflicts on the code and often it is a sign of a missing backport. I don't consider this a big problem. Smaller conflicts (e.g. on the changelog) can be fixed via the github UI.
Often git gets the changelog wrong (adding too many entries). This is not a conflicts, but still require the original author to review and fix the backport.

With mergify doing the backport, will the PR be assigned to the original author?

**Comment by v1v** [comment]
> With mergify doing the backport, will the PR be assigned to the original author?

No, Backport are not assigned to the original contributor. See my comment in https://github.com/elastic/beats/pull/24608#discussion_r597245436

Off the top of my head, an easy `workaround` could be the GH owners, it's not 100% effective but at leats could help with.

Meanwhile, I'll look it at if we can contribute to the mergify source code to provide that feature

**Comment by urso** [comment]
> No, Backport are not assigned to the original contributor. See my comment in #24608 (comment)

Oh, I thaught this is about the author only. As backports still might need manual intervention (changelog), I was just wondering if we can as a fallback just add the original author as 'reviewer' or 'assignee'.
Anyways, the backport should be linked (via the title) to the original PR, so it shouldn't be too hard to find the backport PRs. Don't think it is a blocker.


What is the title of the backport PR? The changelog scripts extract the original PR of a backport based on the PR title.

**Comment by v1v** [comment]
> What is the title of the backport PR? The changelog scripts extract the original PR of a backport based on the PR title.

The tile is the original backport's title in addition to `(bp#<number>)`

A real example, https://github.com/elastic/e2e-testing/pull/935 was the original PR and created:
- https://github.com/elastic/e2e-testing/pull/936
- https://github.com/elastic/e2e-testing/pull/937
- https://github.com/elastic/e2e-testing/pull/938

**Comment by urso** [comment]
> The tile is the original backport's title in addition to (bp#<number>)

Thanks. This format will not match our regex, but it should just be a [one liner here](https://github.com/elastic/ingest-dev/blob/b97b15c1762f7c9fc5ebf1b47fabac8bf0b2aa58/release_scripts/beats-changelog/beats_changelog/bin/__main__.py#L500).

File: .mergify.yml
Comment (urso) [inline .mergify.yml:13]: How exactly will the labels work?

a. If we are about to release 7.12.0, I might want to use the label `v7.12.1`, in order to postpone the backport for after the release is out. Will we update the version filter here after a release? If so, will my PR automatically backported when we update the label filter here?

b. Will my PR be backported if I add `v7.12.0` even after lets say 7.12.1 was released? (Although it might be a bit confusing to users looking at the PR labels to figure out if a bug fix is available)

Comment (v1v) [reply .mergify.yml:13]: I'll answer your questions in a minute, for clarity, the initial proposal was to use the existing labels `vmajor.minor.0` (see https://github.com/elastic/beats/pull/24576#issuecomment-801099897) 

> a. If we are about to release 7.12.0, I might want to use the label v7.12.1, in order to postpone the backport for after the release is out. Will we update the version filter here after a release? If so, will my PR automatically backported when we update the label filter here?

Yes, `mergify` works for even PRs which were not backported after the merge, as long as the label is added. In addition you can use the `github comment` = `@Mergifyio backport <destination> will backport this PR on <destination> branch`

To illustrate the above, I just tagged https://github.com/elastic/apm-pipeline-library/pull/1053 with the explicit label and https://github.com/elastic/apm-pipeline-library/pull/1056 was created. NOTE: for some reason the timestamp for the label does not match with when I did it, (about 7 minutes ago instead of 2 days...)

> b. Will my PR be backported if I add v7.12.0 even after lets say 7.12.1 was released? (Although it might be a bit confusing to users looking at the PR labels to figure out if a bug fix is available)

Yes, backports don't know about releases but branches and tags.

Comment (kuisathaverat) [reply .mergify.yml:13]: >a. If we are about to release 7.12.0, I might want to use the label v7.12.1, in order to postpone the backport for after the release is out. Will we update the version filter here after a release? If so, will my PR automatically backported when we update the label filter here?

if the mergify config is updated with that branch before you merge the PR, the PR would backported.

>b. Will my PR be backported if I add v7.12.0 even after lets say 7.12.1 was released? (Although it might be a bit confusing to users looking at the PR labels to figure out if a bug fix is available)

the backport is made on a new PR when you merge the fixed PR, if the labels and mergify configuration are in place when you merge, a new PR is made to every branch that you requested backport with the labels.

There is an alternative way to make the backports manually using a GitHub comment (see  https://docs.mergify.io/commands.html#backport)

https://mergify.io/features/backports

Comment (urso) [reply .mergify.yml:13]: Maybe to clarify case a:
1. mergify only backports if label `v7.12.0` is set
2. I create a PR and merge it
3. I add the label `v7.12.1` to my PR (because of FF I don't want the change be backported yet)
4. Version 7.12.0 gets released
5. After 7.12.0 is released we update the mergify configuration to backport if `v7.12.0` or `v7.12.1` is set.

After step 5, will my old PR be picked up automatically or am I required to remember that I did have a 2 weeks old PR that I must now 'manually' backport by adding a comment to the PR (or remove + add label again).

Case b is a little unfortunate as labels do not necesarrily reflect reality, but well, it's the way it is.

Comment (kuisathaverat) [reply .mergify.yml:13]: >After 7.12.0 is released we update the mergify configuration to backport if v7.12.0 or v7.12.1 is set.

this happens after you merge the PR, mergify does not review old merged PRs, so the backport is not made. In this case, you have to manually backport the PR, either from the PR with a GiHub Comment or as you make now from your laptop.
However, if we detect changes on the mergify configfile can trigger something that reviews all PRs with the new label and make backport mergify comments on those PRs, but this is something not out of the box. We have to see if this use case exists and how many times it happens, then we can evaluate the effort and the value of this process.

Comment (kuisathaverat) [reply .mergify.yml:13]: AFAIK if you fix something in branch 7.12 that has the version 7.12.0 this will be in the 7.12 (the same major.minor different patch) that will contain the version 7.12.1, so in the case you put as an example there is no backport.
I can see a case similar of what you explain between minor versions, but in those case you use the branch `7.x` and then you backport to `7.12`, `7.11`, ... in those cases the branch already exists so the backport is made.

Comment (v1v) [reply .mergify.yml:13]: To help to unblock this 🧵 

Since there is no backport automation in place for this project I see the benefits of enabling it for facilitating the backporting process, but I don't know if it could cause any issues. Although, `mergify` requires certain discipline regarding what labels should be added. 

Do you think if we enable `mergify` it will help with the backporting? Or is it required to solve all of the above-mentioned use cases?

Comment (urso) [reply .mergify.yml:13]: I don't think either behavior is a blocker and adding mergify is an improvement either way. As we have to document the changelog process in our how to guides. I'm trying to figure out how it works and where we have potential gotchas that we should document as well. I did just bring up the two cases as we've seen them from time to time in the past: label does not match actual release + us holding of with backport in order to target the next patch release... cough... kafka ... cough.

**Review by urso — APPROVED** [review]

**Comment by v1v** [comment]
> With mergify doing the backport, will the PR be assigned to the original author?

I contributed with a new feat -> https://github.com/Mergifyio/mergify-engine/pull/2336 (wait for their feedback)

**Comment by urso** [comment]
> I contributed with a new feat -> Mergifyio/mergify-engine#2336 (wait for their feedback)

Nice. Let's push this PR here anyways. We can still enhance the experience once the mergify feature is merged.

**Comment by v1v** [comment]
If we agree in our monthly meeting today, then I'll proceed with merging this PR.

File: .mergify.yml
Comment (sqren) [inline .mergify.yml:38]: Wouldn't it be easier to ask the user to use a tool for this? Either running mergify locally or [backport](https://github.com/sqren/backport).

In Kibana we help the user like this:
image

Comment (v1v) [reply .mergify.yml:38]: good point. I don't know what's the tool the team have in place for that, in order to be be tool-agnostic the message provides the git commands to solve it locally. 

I'll keep a note to iterate on this and make the user experience smooth. Thanks for the feedback :)

Comment (urso) [reply .mergify.yml:38]: Most people use the `dev-tools/cherry-pick-pr` tool in the Beats repo it seems. But a few contributors use the nodejs backport tool. Both are supported. For the changelog we have a 'requirement' on the PR title to properly link to the original PR, so we can find unique PRs and double check if these features are really released (github tags can be missleading). Both tools are supported by our changelog scripts. Using custom git commands might be a problem if the title is not properly formatted (unless we add another lint check on the PR title for backports).

Comment (v1v) [reply .mergify.yml:38]: We can then iterate with the existing tools, the current comment intends to help with fixing the just created backport PR that has got some conflicts, so we might need to think about that particular use case. I'll keep a note for the upcoming iterations

File: .mergify.yml
Comment (v1v) [inline .mergify.yml:6]:
```suggestion
      - label=backport-v7.13.0
```

File: .mergify.yml
Comment (v1v) [inline .mergify.yml:15]:
```suggestion
      - label=backport-v7.12.0
```

File: .mergify.yml
Comment (v1v) [inline .mergify.yml:24]:
```suggestion
      - label=backport-v7.11.0
```

**Review by v1v — COMMENTED** [review]
As agreed we will roll out this automation in two phases:
- experimental one by using the `backport-vMajor.Minor.0` labels. Therefore, users should use those new tags.
- production, so remove the `backport-vMajor.Minor.0` labels and use the usual ones.

I'll proceed with an email in the mailing list to explain how this will work.

**Pull request merged**
